### PR TITLE
Improve scripts and instructions for macOS & various minor fixes

### PR DESCRIPTION
**Pull request by dime10**
List of changes:
- `tensorflow-cpu` doesn't exist for macOS, use regular `tensorflow` instead in `requirements.txt`
- add instructions to the documentation for building the docs
- fix the default mhlo build directory not being ignored by git (now `bazel-build`, as it has to be a directory already present in the upstream `.gitignore`)
- fix an AutoGraph lit test which produces slightly different output with different TF versions when assigning tuples
- update formatting exclusions (e.g. with Enzyme directories)
- fix test that didn't clean up after itself
- update wheel build recipe with cleanup
- fix extra header in the Catalyst documentation landing page
- add other optional test dependencies to `requirements.txt`
- update installation guide with additional information on building backends and running our various test suites

Otherwise I can confirm that the current build from source instructions work for macOS, as well as all the default Make recipes (build, tests, docs, formatting, demos, wheel). Also checked additional devices and tests (Kokkos, openqasm).
Sanitized builds are left for a follow-up.

### CODE REVIEW

File: .github/workflows/build-wheel-linux-x86_64.yaml [hunk 1]
```diff
@@ -262,7 +262,8 @@ jobs:
               -DKokkos_ENABLE_SERIAL=ON \
               -DKokkos_ENABLE_OPENMP=ON \
               -DENABLE_WARNINGS=OFF \
-              -DENABLE_OPENQASM=ON
+              -DENABLE_OPENQASM=ON \
+              -DENABLE_OPENMP=ON
```
Comment (maliasadi) [inline .github/workflows/build-wheel-linux-x86_64.yaml:266]: We'll add OMP when packaging our wheels with this lib.
```suggestion
              -DENABLE_OPENQASM=ON \
              -DENABLE_OPENMP=OFF
```

File: Makefile [hunk 2]
```diff
@@ -129,6 +129,8 @@ wheel:
 
 	$(PYTHON) $(MK_DIR)/setup.py bdist_wheel
 
+	rm -r $(MK_DIR)/build
+
```
Comment (maliasadi) [inline Makefile:133]: Why did you add this here instead of `clean` and `clean-all`?

**Review by maliasadi — COMMENTED** [review]
I had a few comments about the wheels but the docs update looks really good! :partying_face:

File: .github/workflows/build-wheel-linux-x86_64.yaml (hunk 1 above)
Comment (dime10) [reply .github/workflows/build-wheel-linux-x86_64.yaml:266]: My goal was only to make implicit flags explicit. I checked the lightning repo, and I believe `ON` is the default there (assuming CMake can find an OpenMP package):
https://github.com/PennyLaneAI/pennylane-lightning/blob/74427014b0e7ddc72c33bbadeda53bd97b70bdd5/CMakeLists.txt#L51-L53

Since we are already turning OpenMP on in the Lightning Kokkos config (with the flag `-DKokkos_ENABLE_OPENMP=ON`, I figured we must have already been building the linux wheels with OpenMP enabled on the qubit device.

What do you think?

File: Makefile (hunk 2 above)
Comment (dime10) [reply Makefile:133]: This build folder is only used temporarily when building the wheel, it is not itself the output of the wheel command. I don't think we need to keep it around once the wheel artifact is produced (it just pollutes the user directory).

File: .github/workflows/build-wheel-linux-x86_64.yaml (hunk 1 above)
Comment (maliasadi) [reply .github/workflows/build-wheel-linux-x86_64.yaml:266]: I like the idea of being explicit with these CMake options. The last time Erick and I checked the default value of OMP in lightning.qubit on Linux was OFF. They maybe updated it in the new release! Given the fact that we're not shipping our wheels with the OMP library on Linux and we haven't had no complain so far, it maybe safe to keep it ON.

File: requirements.txt [hunk 3]
```diff
@@ -19,5 +19,7 @@ pytest-xdist
 pytest-cov
 nbmake
 
-# optional rt dependencies
-tensorflow-cpu
+# optional rt/test dependencies
+tensorflow
+amazon-braket-pennylane-plugin>=1.23.0
+pennylane-lightning[kokkos]
```
Comment (maliasadi) [inline requirements.txt:25]: nwm!
```suggestion
pennylane-lightning-kokkos
```

File: .github/workflows/build-wheel-linux-x86_64.yaml [hunk 4]
```diff
@@ -262,7 +262,8 @@ jobs:
               -DKokkos_ENABLE_SERIAL=ON \
               -DKokkos_ENABLE_OPENMP=ON \
               -DENABLE_WARNINGS=OFF \
-              -DENABLE_OPENQASM=ON
```
Comment (maliasadi) [inline .github/workflows/build-wheel-linux-x86_64.yaml:265]: You can remove line 261 and update line 257 as follows:
```
              -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=$GITHUB_WORKSPACE/runtime-build/lib \
```

Comment (dime10) [reply .github/workflows/build-wheel-linux-x86_64.yaml:265]: Why are we removing the git tag specification?

File: doc/dev/installation.rst [hunk 5]
```diff
@@ -144,28 +144,37 @@ directory:
...
+and remote services, such as Amazon Braket. The full list of supported backends, and additional
+configuration options, are available in the
 `Catalyst Runtime <https://docs.pennylane.ai/projects/catalyst/en/latest/modules/runtime.html>`_
 page.
 
+To install Catalyst with all available backends, simply run:
+
+.. code-block:: console
+
+  make runtime ENABLE_LIGHTNING_KOKKOS=ON ENABLE_OPENQASM=ON
```
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/installation.rst:176]: https://github.com/PennyLaneAI/catalyst/pull/363/commits/f83970b864e24606a4471ba67cf0a6c98ac99d47

File: .github/workflows/build-wheel-linux-x86_64.yaml (hunk 4 above)
Comment (maliasadi) [reply .github/workflows/build-wheel-linux-x86_64.yaml:265]: yeah, I've added it and realized this's not make commands :smile: I thought that I updated my code comment before posting. Anyway, I've edited the suggestion

**Review by maliasadi — APPROVED** [review]
Thanks @dime10! :tada:

File: doc/dev/installation.rst [hunk 6]
```diff
@@ -10,6 +10,14 @@ pre-built binaries are being distributed via the Python Package Index (PyPI) for
 
     pip install pennylane-catalyst
 
+.. warning::
+
+    macOS does not ship with a system compiler by default, which Catalyst depends on. Please
+    ensure that `XCode <https://developer.apple.com/xcode/resources/`_ or the
+    ``XCode Command Line Tools`` are installed on your system before using Catalyst.
```
Comment (maliasadi) [inline doc/dev/installation.rst:17]: Have you tried testing Catalyst with the clang installation from Homebrew? This may be an alternative option here.

Comment (dime10) [reply doc/dev/installation.rst:17]: Haven't tried it but it should be possible. However for regular users installing Xcode is one less step to get Catalyst (no need to install brew).
